### PR TITLE
Parenthesize type assertions when necessary

### LINKED ISSUE #4642: `(<any>{})` (object literal) emits unexpected code `{}` (code block).

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
**Before reviewing this, please review #4914**.

Fixes #4642.

This change fixes emit for a number of cases, but specifically for type assertions when the innermost expression can be parsed differently in an *ExpressionStatement* position.

For instance, take:

```TypeScript
<any>{} + 10;
```

This was originally emitted as

```JavaScript
{};
+10;
```

which is a block body followed by the expression statement `+10`, instead of 

```JavaScript
({} + 10)
```

which is an addition expression with an empty object and `10`.

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -996,6 +996,13 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                 if (node.template.kind === SyntaxKind.TemplateExpression) {
                     forEach((<TemplateExpression>node.template).templateSpans, templateSpan => {
                         write(", ");
+                        // TODO (drosen): Something like
+                        //      `${ { hello: 10; world: 20 } }HELLO`
+                        // will get emitted as
+                        //      { hello: 10; world: 20 } + "HELLO"
+                        //
+                        // As an expresssion statement, the former evaluates to '"[object Object]HELLO"' while the latter evaluates to 'NaN'.
+                        // Look into using 'meaningChangesAsExpressionStatement' here.
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: See #4912.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -2335,46 +2307,130 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                const isBinaryExpr = innerExpr.kind === SyntaxKind.BinaryExpression;
+                if (isBinaryExpr) {
+                    const operator = (<BinaryExpression>innerExpr).operatorToken;
+                    if (operator.kind === SyntaxKind.CommaToken) {
+                        return nodeIsSensitiveToCommaExpressions(originalParent);
+                    }
+
+                    if (comparePrecedenceToBinaryPlus(innerExpr) !== Comparison.GreaterThan) {
+                        let parentIsAddition = originalParentKind === SyntaxKind.BinaryExpression &&
+                            (<BinaryExpression>originalParent).operatorToken.kind === SyntaxKind.PlusToken;
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: This check is insufficient, but we actually don't have tests for this right now. I'll add some soon.

**Comment by mhegazy** [comment]
@DanielRosenwasser can you please refresh.

**Comment by DanielRosenwasser** [comment]
@mhegazy I'm waiting on @yuit's exponentiation work.

**Comment by mhegazy** [comment]
@DanielRosenwasser can you please refresh.

**Pull request closed without merging**